<a href="https://colab.research.google.com/github/mwakitrevor146-prog/machine-learning-projects/blob/main/lasso.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install pandas scikit-learn
import pandas as pd
from sklearn.linear_model import Lasso
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

In [2]:
# Sample dataset: Study hours, previous exam scores, and pass/fail labels
data = {
    'StudyHours': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    'PrevExamScore': [30, 40, 45, 50, 60, 65, 70, 75, 80, 85],
    'Pass': [0, 0, 0, 0, 0, 1, 1, 1, 1, 1]  # 0 = Fail, 1 = Pass
}

df = pd.DataFrame(data)

# Features and target variable
X = df[['StudyHours', 'PrevExamScore']]  # Features
y = df['Pass']  # Target variable


# Split the data into training and testing sets (80% training, 20% testing)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [3]:
# Initialize the LASSO model with alpha (regularization parameter)
lasso_model = Lasso(alpha=0.1)

# Train the model on the training data
lasso_model.fit(X_train, y_train)

# Make predictions on the test data
y_pred = lasso_model.predict(X_test)

# Evaluate the model's performance using R-squared
r2 = r2_score(y_test, y_pred)
print(f'R-squared score: {r2}')

R-squared score: 0.9997884297520662


In [4]:
# Display the coefficients of the features
print(f'LASSO Coefficients: {lasso_model.coef_}')

LASSO Coefficients: [0.         0.02463636]


### Step 1: Library Installation and Import

First, you installed the necessary libraries (`pandas` for data manipulation and `scikit-learn` for machine learning) and then imported the specific modules you'd need: `Lasso` for the regression model, `train_test_split` for dividing your data, and `r2_score` for evaluating the model.

### Step 2: Data Preparation

You created a sample dataset with 'StudyHours', 'PrevExamScore' as features, and 'Pass' as the target variable. This data was then converted into a pandas DataFrame. Following this, you split the data into training and testing sets. This is a crucial step in machine learning to evaluate how well your model generalizes to unseen data. `test_size=0.2` means 20% of the data was reserved for testing, and `random_state=42` ensures reproducibility of the split.

### Step 3: Model Training and Prediction

You initialized a `Lasso` regression model with an `alpha` (regularization parameter) of 0.1. Lasso (Least Absolute Shrinkage and Selection Operator) is a type of linear regression that uses shrinkage. Shrinkage is where data values are shrunk towards a central point, like the mean. This helps to prevent overfitting and can perform feature selection by forcing some coefficients to become exactly zero.

The model was then trained using your training data (`X_train`, `y_train`), and predictions (`y_pred`) were made on the test data (`X_test`).

### Step 4: Model Evaluation and Coefficient Analysis

The `r2_score` (R-squared score) was used to evaluate the model's performance. An R-squared score of `0.9997884297520662` is extremely high, indicating that nearly all the variance in the target variable (`Pass`) can be explained by your features (`StudyHours`, `PrevExamScore`) within this small dataset.

The LASSO coefficients are `[0. 0.02463636]`. This indicates:
- The coefficient for 'StudyHours' is `0.0`, meaning LASSO has completely 'shrunk' this feature's coefficient to zero. This implies that 'StudyHours' is not considered a significant predictor by the model given the `alpha` value and the presence of 'PrevExamScore'.
- The coefficient for 'PrevExamScore' is approximately `0.0246`. This suggests that for every one-unit increase in 'PrevExamScore', the predicted 'Pass' outcome (which is a continuous value in Lasso regression, even though your original 'Pass' is binary) increases by `0.0246`, assuming 'StudyHours' remains constant.